File: sdk/python/stellar_insured_sdk/schemas.py
"""Pydantic models for schema-validated contract payloads and result decoding."""

from __future__ import annotations

from enum import Enum
from typing import Annotated

from pydantic import BaseModel, Field, field_validator

# ── Common validators ─────────────────────────────────────────────────────────

StellarAddress = Annotated[str, Field(min_length=1, pattern=r"^[A-Z0-9]+$")]
PositiveInt = Annotated[int, Field(gt=0)]
NonNegativeInt = Annotated[int, Field(ge=0)]
Percentage = Annotated[int, Field(ge=0, le=100)]
BasisPoints = Annotated[int, Field(ge=0, le=10000)]


# ── Policy ────────────────────────────────────────────────────────────────────


class PolicyStatus(str, Enum):
    ACTIVE = "Active"
    EXPIRED = "Expired"
    CANCELLED = "Cancelled"


class PolicyRecord(BaseModel):
    coverage: PositiveInt
    premium: PositiveInt
    holder: StellarAddress
    status: PolicyStatus
    issue_date: NonNegativeInt
    expiry_date: NonNegativeInt
    total_fractions: NonNegativeInt


class InitializePolicyParams(BaseModel):
    admin: StellarAddress
    guardian: StellarAddress


class IssuePolicyParams(BaseModel):
    holder: StellarAddress
    policy_id: NonNegativeInt
    coverage: PositiveInt
    premium: PositiveInt


class IssuePolicyWithDurationParams(IssuePolicyParams):
    duration_days: PositiveInt


class CalculateDynamicPremiumParams(BaseModel):
    base_premium: PositiveInt
    risk_score: Annotated[int, Field(ge=0, le=100)]
    location_factor: NonNegativeInt
    coverage_ratio: NonNegativeInt


class CreateAmmPoolParams(BaseModel):
    policy_id: NonNegativeInt
    stable_amount: PositiveInt


class SwapPolicyFractionParams(BaseModel):
    buyer: StellarAddress
    policy_id: NonNegativeInt
    stable_in: PositiveInt


class RequestEndorsementParams(BaseModel):
    requester: StellarAddress
    policy_id: NonNegativeInt
    new_coverage: PositiveInt
    new_premium: PositiveInt
    reason: Annotated[str, Field(min_length=1)]


class ApproveEndorsementParams(BaseModel):
    caller: StellarAddress
    endorsement_id: NonNegativeInt


class RejectEndorsementParams(BaseModel):
    caller: StellarAddress
    endorsement_id: NonNegativeInt
    reason: Annotated[str, Field(min_length=1)]


# ── Claims ────────────────────────────────────────────────────────────────────


class ClaimStatus(str, Enum):
    PENDING = "Pending"
    APPROVED = "Approved"
    REJECTED = "Rejected"
    SETTLED = "Settled"


class FraudStatus(str, Enum):
    CLEAN = "Clean"
    SUSPICIOUS = "Suspicious"
    CONFIRMED = "Confirmed"


class PayoutStatus(str, Enum):
    PENDING = "Pending"
    COMPLETED = "Completed"
    FAILED = "Failed"


class ClaimRecord(BaseModel):
    policy_id: NonNegativeInt
    amount: PositiveInt
    status: ClaimStatus
    claimant: StellarAddress
    evidence_count: NonNegativeInt
    fraud_score: NonNegativeInt


class PayoutRecord(BaseModel):
    claim_id: NonNegativeInt
    recipient: StellarAddress
    amount: PositiveInt
    status: PayoutStatus
    retry_count: NonNegativeInt


class SubmitClaimParams(BaseModel):
    policy_address: StellarAddress
    claim_id: NonNegativeInt
    policy_id: NonNegativeInt
    amount: PositiveInt


class SubmitEvidenceParams(BaseModel):
    claim_id: NonNegativeInt
    ipfs_hash: Annotated[str, Field(min_length=1)]
    sensitive: bool
    description: Annotated[str, Field(min_length=1)]
    submitter: StellarAddress


class VerifyEvidenceParams(BaseModel):
    caller: StellarAddress
    evidence_id: NonNegativeInt
    is_valid: bool
    notes: str


class FlagClaimParams(BaseModel):
    caller: StellarAddress
    claim_id: NonNegativeInt
    score_adjustment: int


class SetPayoutTokenParams(BaseModel):
    caller: StellarAddress
    token_address: StellarAddress


# ── Risk Pool ─────────────────────────────────────────────────────────────────


class LiquidityProvider(BaseModel):
    deposited: NonNegativeInt
    allocated_rewards: NonNegativeInt
    claimed_rewards: NonNegativeInt
    vesting_start: NonNegativeInt


class VestingConfig(BaseModel):
    cliff_period: NonNegativeInt
    duration: NonNegativeInt
    early_withdrawal_penalty_bps: BasisPoints


class DepositParams(BaseModel):
    sender: StellarAddress = Field(alias="from")
    amount: PositiveInt


class WithdrawParams(BaseModel):
    to: StellarAddress
    amount: PositiveInt


class SetVestingParams(BaseModel):
    caller: StellarAddress
    cliff_secs: NonNegativeInt
    duration_secs: PositiveInt
    penalty_bps: BasisPoints


class AllocateRewardsParams(BaseModel):
    caller: StellarAddress
    provider: StellarAddress
    amount: PositiveInt


class AppealSlashingParams(BaseModel):
    appealer: StellarAddress
    claim_id: NonNegativeInt
    deposit: PositiveInt
    slashed_amount: PositiveInt


class ResolveAppealParams(BaseModel):
    caller: StellarAddress
    claim_id: NonNegativeInt
    approved: bool
    refund_percentage: Percentage


class SetReinsuranceParams(BaseModel):
    caller: StellarAddress
    reinsurer: StellarAddress
    percentage: Percentage
    credit_score: NonNegativeInt


# ── Governance ────────────────────────────────────────────────────────────────


class ProposalStatus(str, Enum):
    ACTIVE = "Active"
    PASSED = "Passed"
    REJECTED = "Rejected"
    EXECUTED = "Executed"
    COMMITTED = "Committed"


class Proposal(BaseModel):
    id: NonNegativeInt
    proposer: StellarAddress
    title: str
    description: str
    status: ProposalStatus
    voting_deadline: NonNegativeInt


class InitializeGovernanceParams(BaseModel):
    admin: StellarAddress
    token_contract: StellarAddress
    voting_period_days: PositiveInt
    min_voting_percentage: Percentage
    min_quorum_percentage: Percentage
    slashing_contract: StellarAddress


class CreateProposalParams(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=256)]
    description: Annotated[str, Field(min_length=1)]
    execution_data: str
    threshold_percentage: Percentage


class VoteParams(BaseModel):
    proposal_id: NonNegativeInt
    vote_weight: PositiveInt
    is_yes: bool


class CreateSlashingProposalParams(BaseModel):
    target: StellarAddress
    role: Annotated[str, Field(min_length=1)]
    reason: Annotated[str, Field(min_length=1)]
    amount: PositiveInt
    evidence: str
    threshold: Percentage


# ── Shared ────────────────────────────────────────────────────────────────────


class SetPauseStateParams(BaseModel):
    caller: StellarAddress
    is_paused: bool
    reason: Annotated[str, Field(min_length=1)]

File: sdk/typescript/src/schemas.ts
import { z } from "zod";

// ─── Common ──────────────────────────────────────────────────────────────────

const SorobanAddress = z
  .string()
  .min(1, "Address must not be empty")
  .regex(/^[A-Z0-9]+$/, "Invalid Stellar address format");

const PositiveU64 = z.coerce.bigint().refine((v) => v > 0n, {
  message: "Value must be a positive integer",
});

const NonNegativeU64 = z.coerce.bigint().refine((v) => v >= 0n, {
  message: "Value must be non-negative",
});

const PolicyId = z.coerce.bigint().refine((v) => v >= 0n, {
  message: "Policy ID must be non-negative",
});

const ClaimId = z.coerce.bigint().refine((v) => v >= 0n, {
  message: "Claim ID must be non-negative",
});

// ─── Policy Contract ─────────────────────────────────────────────────────────

export const PolicyStatusSchema = z.enum([
  "Active",
  "Expired",
  "Cancelled",
]);

export const PolicyRecordSchema = z.object({
  coverage: PositiveU64,
  premium: PositiveU64,
  holder: SorobanAddress,
  status: PolicyStatusSchema,
  issue_date: NonNegativeU64,
  expiry_date: NonNegativeU64,
  total_fractions: NonNegativeU64,
});

export const InitializePolicySchema = z.object({
  admin: SorobanAddress,
  guardian: SorobanAddress,
});

export const IssuePolicySchema = z.object({
  holder: SorobanAddress,
  policy_id: PolicyId,
  coverage: PositiveU64,
  premium: PositiveU64,
});

export const IssuePolicyWithDurationSchema = IssuePolicySchema.extend({
  duration_days: z.number().int().positive("Duration must be positive"),
});

export const CancelPolicySchema = z.object({
  policy_id: PolicyId,
});

export const ExpirePolicySchema = z.object({
  policy_id: PolicyId,
});

export const CheckExpirePoliciesSchema = z.object({
  start_index: NonNegativeU64,
  max_items: PositiveU64,
});

export const CalculateDynamicPremiumSchema = z.object({
  base_premium: PositiveU64,
  risk_score: z.number().int().min(0).max(100),
  location_factor: z.number().int().min(0),
  coverage_ratio: z.number().int().min(0),
});

export const CreateAmmPoolSchema = z.object({
  policy_id: PolicyId,
  stable_amount: PositiveU64,
});

export const SwapPolicyFractionSchema = z.object({
  buyer: SorobanAddress,
  policy_id: PolicyId,
  stable_in: PositiveU64,
});

export const RequestEndorsementSchema = z.object({
  requester: SorobanAddress,
  policy_id: PolicyId,
  new_coverage: PositiveU64,
  new_premium: PositiveU64,
  reason: z.string().min(1, "Reason must not be empty"),
});

export const ApproveEndorsementSchema = z.object({
  caller: SorobanAddress,
  endorsement_id: NonNegativeU64,
});

export const RejectEndorsementSchema = z.object({
  caller: SorobanAddress,
  endorsement_id: NonNegativeU64,
  reason: z.string().min(1, "Reason must not be empty"),
});

// ─── Claims Contract ─────────────────────────────────────────────────────────

export const ClaimStatusSchema = z.enum([
  "Pending",
  "Approved",
  "Rejected",
  "Settled",
]);

export const FraudStatusSchema = z.enum([
  "Clean",
  "Suspicious",
  "Confirmed",
]);

export const PayoutStatusSchema = z.enum([
  "Pending",
  "Completed",
  "Failed",
]);

export const ClaimRecordSchema = z.object({
  policy_id: PolicyId,
  amount: PositiveU64,
  status: ClaimStatusSchema,
  claimant: SorobanAddress,
  evidence_count: NonNegativeU64,
  fraud_score: z.number().int().min(0),
});

export const PayoutRecordSchema = z.object({
  claim_id: ClaimId,
  recipient: SorobanAddress,
  amount: PositiveU64,
  status: PayoutStatusSchema,
  retry_count: z.number().int().min(0),
});

export const InitializeClaimsSchema = z.object({
  admin: SorobanAddress,
  guardian: SorobanAddress,
});

export const SubmitClaimSchema = z.object({
  policy_address: SorobanAddress,
  claim_id: ClaimId,
  policy_id: PolicyId,
  amount: PositiveU64,
});

export const SubmitEvidenceSchema = z.object({
  claim_id: ClaimId,
  ipfs_hash: z.string().min(1, "IPFS hash required"),
  sensitive: z.boolean(),
  description: z.string().min(1, "Description required"),
  submitter: SorobanAddress,
});

export const VerifyEvidenceSchema = z.object({
  caller: SorobanAddress,
  evidence_id: NonNegativeU64,
  is_valid: z.boolean(),
  notes: z.string(),
});

export const FlagClaimSchema = z.object({
  caller: SorobanAddress,
  claim_id: ClaimId,
  score_adjustment: z.number().int(),
});

export const SetPayoutTokenSchema = z.object({
  caller: SorobanAddress,
  token_address: SorobanAddress,
});

// ─── Risk Pool Contract ──────────────────────────────────────────────────────

export const LiquidityProviderSchema = z.object({
  deposited: NonNegativeU64,
  allocated_rewards: NonNegativeU64,
  claimed_rewards: NonNegativeU64,
  vesting_start: NonNegativeU64,
});

export const VestingConfigSchema = z.object({
  cliff_period: NonNegativeU64,
  duration: NonNegativeU64,
  early_withdrawal_penalty_bps: z.number().int().min(0).max(10000),
});

export const InitializeRiskPoolSchema = z.object({
  admin: SorobanAddress,
  guardian: SorobanAddress,
});

export const DepositSchema = z.object({
  from: SorobanAddress,
  amount: PositiveU64,
});

export const WithdrawSchema = z.object({
  to: SorobanAddress,
  amount: PositiveU64,
});

export const SetVestingSchema = z.object({
  caller: SorobanAddress,
  cliff_secs: NonNegativeU64,
  duration_secs: PositiveU64,
  penalty_bps: z.number().int().min(0).max(10000),
});

export const AllocateRewardsSchema = z.object({
  caller: SorobanAddress,
  provider: SorobanAddress,
  amount: PositiveU64,
});

export const AppealSlashingSchema = z.object({
  appealer: SorobanAddress,
  claim_id: ClaimId,
  deposit: PositiveU64,
  slashed_amount: PositiveU64,
});

export const ResolveAppealSchema = z.object({
  caller: SorobanAddress,
  claim_id: ClaimId,
  approved: z.boolean(),
  refund_percentage: z.number().int().min(0).max(100),
});

export const SetReinsuranceSchema = z.object({
  caller: SorobanAddress,
  reinsurer: SorobanAddress,
  percentage: z.number().int().min(0).max(100),
  credit_score: z.number().int().min(0),
});

// ─── Governance Contract ─────────────────────────────────────────────────────

export const ProposalStatusSchema = z.enum([
  "Active",
  "Passed",
  "Rejected",
  "Executed",
  "Committed",
]);

export const ProposalSchema = z.object({
  id: NonNegativeU64,
  proposer: SorobanAddress,
  title: z.string(),
  description: z.string(),
  status: ProposalStatusSchema,
  voting_deadline: NonNegativeU64,
});

export const InitializeGovernanceSchema = z.object({
  admin: SorobanAddress,
  token_contract: SorobanAddress,
  voting_period_days: z.number().int().positive(),
  min_voting_percentage: z.number().int().min(0).max(100),
  min_quorum_percentage: z.number().int().min(0).max(100),
  slashing_contract: SorobanAddress,
});

export const CreateProposalSchema = z.object({
  title: z.string().min(1, "Title required").max(256, "Title too long"),
  description: z.string().min(1, "Description required"),
  execution_data: z.string(),
  threshold_percentage: z.number().int().min(0).max(100),
});

export const VoteSchema = z.object({
  proposal_id: NonNegativeU64,
  vote_weight: PositiveU64,
  is_yes: z.boolean(),
});

export const FinalizeProposalSchema = z.object({
  proposal_id: NonNegativeU64,
});

export const CreateSlashingProposalSchema = z.object({
  target: SorobanAddress,
  role: z.string().min(1),
  reason: z.string().min(1),
  amount: PositiveU64,
  evidence: z.string(),
  threshold: z.number().int().min(0).max(100),
});

// ─── Pause State (shared) ────────────────────────────────────────────────────

export const SetPauseStateSchema = z.object({
  caller: SorobanAddress,
  is_paused: z.boolean(),
  reason: z.string().min(1, "Reason required"),
});

// ─── Type exports ────────────────────────────────────────────────────────────

export type PolicyRecord = z.infer<typeof PolicyRecordSchema>;
export type ClaimRecord = z.infer<typeof ClaimRecordSchema>;
export type PayoutRecord = z.infer<typeof PayoutRecordSchema>;
export type LiquidityProvider = z.infer<typeof LiquidityProviderSchema>;
export type VestingConfig = z.infer<typeof VestingConfigSchema>;
export type Proposal = z.infer<typeof ProposalSchema>;
